src/View/Helper/PaginatorHelper.php

```diff
@@ -792,17 +792,28 @@ public function numbers(array $options = [])
     protected function _getNumbersStartAndEnd($params, $options)
     {
         $half = (int)($options['modulus'] / 2);
-        $end = $params['page'] + $half;
+        $end = max(1 + $options['modulus'], $params['page'] + $half);
+        $start = min($params['pageCount'] - $options['modulus'], $params['page'] - $half - $options['modulus'] % 2);
 
-        if ($end > $params['pageCount']) {
-            $end = $params['pageCount'];
+        if ($options['first']) {
+            $first = is_int($options['first']) ? $options['first'] : 1;
+
+            if ($start <= $first + 2) {
```

Comment:
Numbers starts from page `1` + number of first links + `1` ellipsis. The `1+1` is merged into 2.

This is reverse of `$end >= $params['pageCount'] - $last - 1`, where `$params['pageCount']` is last page (ex. 20) - number of last links - 1 ellipsis.

Example:

If 'first' is 3, then numbers should begin from 1, when `$start <= 1+3+1`, because ellipsis in `1 2 3 ... 5 6 7` skips only 1 page link, so ellipsis is useless and it is better to display link to 4th page.

Ellipsis only needed when `$start` is 6 or more: `1 2 3 ... 6 7 8`.